### DISCUSSION TITLE
[@types/better-sqlite3]

### DISCUSSION THREAD

**Discussion by AverageHelper** [DefinitelyTyped/DefinitelyTyped#62205] [Issues with a @types package] [answered]
Trying to continue the discussion in #50794. I have [a fork](https://github.com/AverageHelper/DefinitelyTyped/blob/avg/better-sqlite3-prefer-unknown-to-any/types/better-sqlite3/index.d.ts) ready, with type tests done as well.

I made this change for a friend who no longer uses `better-sqlite3`, so I have no easy way to test this. I'm posting here to ask @leafac if this is what they were looking to implement, and whether someone has a script that might exercise these types enough to make a PR viable here.

[a comment by an automated account is not shown]

**Comment by leafac** [comment]
@AverageHelper

Thank you for that. I’ll take a look shortly… 😁

**Comment by leafac** [comment]
@AverageHelper I had in mind something different from what you did. In several places you changed from `any` to `unknown`, while I was thinking on introducing generics instead.

For example, you turned `get(...params: BindParameters): any;` into `get(...params: BindParameters): unknown;`, while I would have done something like `get<T>(...params: BindParameters): T;`.

What do you think of this?

Did you try and it didn’t work out for some reason?

Thanks again for working on this 😀

**Reply by AverageHelper** [reply]
Thank you for looking this over!

I considered using a generic type parameter there. I decided against the generic because, since the function parameters do not use the type argument, the generic is essentially an implicit type cast.

TL;DR: `unknown` is much safer in this case.

```ts
// For example, consider this type and a simple type guard:
interface Something {
	a: boolean;
	b: string;
}

function assertSomething(tbd: unknown): asserts tbd is Something;

// Using generics, the implicit type cast is clobbered:
function get<T>(): T;
const r1: Something = get<Something>(); // unsafe, as TS assumes without question that the value returned is a `Something`
```

Since this data is from a database—and is therefore a sort of API boundary—a runtime check should be preferred, for safety. But because the call accepts a generic parameter, most callers will avoid using `unknown` there, instead writing the type they expect to receive. Because TypeScript is placated, callers will likely forget to check the value at runtime, leading to potential trouble later on, and defeating the point of TypeScript.

```ts
// Using `unknown`, TS forces a choice, either a type guard or explicit type cast:
function get(): unknown;
const r2: Something = get() as Something; // unsafe, but the caller knows this

// Now if the caller wants to be safer, they may do this instead:
const r3 = get(); // r3 is of type `unknown`
assertSomething(r3); // Throws if r3 is unexpectedly not a `Something`, so hereafter r3 is of type `Something`
console.log(r3.b); // No compile-time error, and no runtime `TypeError` is possible!
```

In the case of the `get` function in question, there isn't any more type safety between the generic form and casting `unknown` explicitly, using `as`. The only difference is where the manual type cast assertion goes, and how careful each method forces the caller to be. TypeScript in both cases cannot verify that the result is actually of the given type without some work at runtime.

If the caller is _certain_ that their data is always well-formed, then a basic type cast is fine. If not, then TypeScript should speak up about the potentially-dangerous use of unsanitized data. A compile-time typecast or assertion should be very clear, using `as`. Otherwise, a type guard (or other runtime handling of bad data) should be required.

If, instead, the function's parameters used the generic type argument in some way, whether by taking an actual instance of a case of `T`, or taking a type guard function that asserts or checks some case of `T`, or some other way that can be statically analyzed and assured, then a generic argument would be preferable to `unknown`.

Here, I prefer `unknown`, to make clear to callers that they must either (1) manually check at runtime that the data they're getting is what they expect it to be, or else (2) tell TypeScript "trust me bro" with an explicit type assertion.

**Reply by AverageHelper** [reply]
Looking back at the code, it seems that the `prepare` method of the `Database` interface also falls for the above-described pitfall (generic type parameter, but only using a `string` arg at runtime). I left that method mostly alone. Since I haven't used that method yet, I'm not sure the best form that signature should take.

**Reply by AverageHelper** [reply]
If I am wrong in my reasoning, please inform me! I'm always learning about TypeScript, and how I can use it more effectively

**Reply by leafac** [reply, marked as answer]
@AverageHelper

You explained the tradeoffs remarkably well. Thank you for the detailed explanation.

I agree with you that `unknown` is better than generics. Reasons:

1. I proposed generics to begin with because that’s how [`sqlite` does it](https://github.com/kriasoft/node-sqlite/blob/655fa256f9c6390debb9d3251ea6a0ce25324fd0/src/Statement.ts#L119).

2. `.get<{ ... }>()` and `.get() as { ... }` seem equally ergonomic to me.

3. You’re right that people who want runtime checks (I’m not one of them 🤷) will be nudged in the right direction with the use of `unknown`.

I say: Let’s go for it 👍

**Reply by AverageHelper** [reply]
Then I'll get started on that PR! Thank you very much for looking this over

**Comment by AverageHelper** [comment]
Approved and merged in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65035!
